### ISSUE TITLE
FaceLocator

### ISSUE THREAD

**Issue by johndpope** [johndpope/Emote-hack#8]
The EMO paper describes the Face Locator as a component designed to work with the generative model to control where the character's face is generated in the output video. Here's a detailed breakdown based on the descriptions provided in the excerpts:

Weak Control Signal: The Face Locator is part of a "weak" control signal approach that identifies the facial region within the frame. This region encompasses the bounding box of the face in the video clip.

Component Design: The Face Locator comprises lightweight convolutional layers that encode the bounding box mask. The resulting encoded mask is then used during the generation process in the Backbone Network.

Functionality:

The Face Locator takes an image as input and identifies the facial region.
It encodes the location of the face as a mask, which can then be used to ensure that the face is generated within the correct region of the latent representation.
Integration with Generative Process: Once the facial bounding box mask is encoded by the Face Locator, it is added to the noisy latent representation before being fed into the Backbone Network. This method allows the model to control the facial region generation during the video synthesis process.

Ensuring Stability and Consistency: The Face Locator works alongside other model components, such as temporal modules and speed layers, to ensure that the generated character’s motion is consistent and stable across video clips. This is crucial because independent generation processes might otherwise result in inconsistent motion.

Facial Region as Weak Control: The specified face region by the Face Locator does not constitute strong control conditions. Instead, it indicates a sizeable area within which facial movement is permissible, ensuring the head is not restricted to a static posture.

Complementing Speed Layers: The Face Locator's output is used in conjunction with speed layers that modulate the head rotation speed, contributing to the production of smooth and controlled motion.

The actual code implementation of the Face Locator would involve using a pre-trained facial detection model to identify the bounding box of the face and encoding this information into a form that can be utilized by the generative model to control face synthesis. The bounding box is typically represented as a mask that highlights the facial region, which is then passed as a conditioning input to the generative model, influencing where facial features appear in the generated frames.

FaceMeshGenerator.py - seems ok. 
```python 

 def generate_mask(self, image):

        # Convert image to RGB
        image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

        # Create a blank mask with the same dimensions as the image
        mask = np.zeros(image.shape[:2], dtype=np.uint8)

        # Detect faces
        with self.mp_face_detection.FaceDetection(model_selection=1, min_detection_confidence=0.5) as face_detection:
            detection_results = face_detection.process(image_rgb)

        # If faces are detected, find the face landmarks.
        if detection_results.detections:
            with self.mp_face_mesh.FaceMesh(
                static_image_mode=True,
                max_num_faces=1,
                min_detection_confidence=0.5) as face_mesh:
                
                # Apply the face mesh model.
                mesh_results = face_mesh.process(image_rgb)
                
                if mesh_results.multi_face_landmarks:
                    for face_landmarks in mesh_results.multi_face_landmarks:
                        # Draw the face mesh on the mask
                        for landmark in face_landmarks.landmark:
                            x = min(int(landmark.x * image.shape[1]), image.shape[1] - 1)
                            y = min(int(landmark.y * image.shape[0]), image.shape[0] - 1)
                            mask[y, x] = 255

        # Optional: Expand the mask to cover a larger area if needed
        # For example, you can use cv2.dilate() to expand the mask area

        # Convert the mask to a PyTorch tensor
        mask_tensor = torch.tensor(mask, dtype=torch.float32)
        return mask_tensor
```

**Comment by johndpope** [comment]
![Screenshot from 2024-03-13 10-32-48](https://github.com/johndpope/Emote-hack/assets/289994/1c5502b8-9198-4bd0-a653-13431fc0a541)

https://github.com/johndpope/Emote-hack/blob/main/FaceLocator.py

need clarity on how to works alongside other model components, such as temporal modules and speed layers, to ensure that the generated character’s motion is consistent and stable across video clips

**Comment by johndpope** [comment]
i create this - we don't actually need this i think - https://github.com/johndpope/Emote-hack/blob/main/train_stage_0.py
